src/librustc_typeck/check/method/probe.rs

```diff
@@ -594,25 +631,37 @@ impl<'a, 'gcx, 'tcx> ProbeContext<'a, 'gcx, 'tcx> {
         Ok(())
     }
 
+    pub fn matches_return_type(&self, method: &ty::ImplOrTraitItem<'tcx>,
+                               expected: ty::Ty<'tcx>) -> bool {
+        match *method {
+            ty::ImplOrTraitItem::MethodTraitItem(ref x) => {
+                self.probe(|_| {
+                    let output = self.replace_late_bound_regions_with_fresh_var(
+                        self.span, infer::FnCall, &x.fty.sig.output());
+                    let substs = self.fresh_substs_for_item(self.span, method.def_id());
+                    let output = output.0.subst(self.tcx, substs);
```

Comment:
Nit: It doesn't really make any difference, but normally we would substitute early-bound stuff first, and I think i'd prefer to follow the usual ordering here. Basically `let output = x.fty.sig.output().subst(self.tcx, substs)`, and then `let (output, _) = self.replace_late_bound(...)`